org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/HoverInfoProvider.java

```diff
@@ -86,21 +95,53 @@ public HoverInfoProvider(ITypeRoot aUnit) {
 			} else {
 				curr = elements[0];
 			}
-			MarkedString signature = this.computeSignature(curr);
-			if (signature != null) {
-				res.add(Either.forRight(signature));
-			}
-			String javadoc = computeJavadocHover(curr);
-			if (javadoc != null) {
-				res.add(Either.forLeft(javadoc));
+			boolean resolved = isResolved(curr);
+			if (resolved) {
+				MarkedString signature = this.computeSignature(curr);
+				if (signature != null) {
+					res.add(Either.forRight(signature));
+				}
+				String javadoc = computeJavadocHover(curr);
+				if (javadoc != null) {
+					res.add(Either.forLeft(javadoc));
+				}
 			}
-
 		} catch (Exception e) {
 			JavaLanguageServerPlugin.logException("Error computing hover", e);
 		}
 		return res;
 	}
 
+	private boolean isResolved(IJavaElement element) throws CoreException {
+		if (element.getElementType() != IJavaElement.TYPE) {
+			return true;
+		}
+		SearchPattern pattern = SearchPattern.createPattern(element, IJavaSearchConstants.ALL_OCCURRENCES);
+		final boolean[] res = new boolean[1];
+		res[0] = false;
+		SearchEngine engine = new SearchEngine();
+		IJavaSearchScope scope = SearchEngine.createJavaSearchScope(new IJavaElement[] { unit }, IJavaSearchScope.SOURCES);
+		engine.search(pattern, new SearchParticipant[] { SearchEngine.getDefaultSearchParticipant() }, scope, new SearchRequestor() {
+
+			@Override
+			public void acceptSearchMatch(SearchMatch match) throws CoreException {
+				if (match.getAccuracy() == SearchMatch.A_INACCURATE) {
+					return;
+				}
+				Object o = match.getElement();
+				if (o instanceof IJavaElement) {
+					IJavaElement element = (IJavaElement) o;
+					ICompilationUnit compilationUnit = (ICompilationUnit) element.getAncestor(IJavaElement.COMPILATION_UNIT);
+					if (compilationUnit == null) {
+						return;
+					}
+					res[0] = true;
+				}
+			}
+		}, new NullProgressMonitor());
```

Comment:
we should get the monitor from the initial request to support cancellation. That means changing a bunch of methods signatures along the way.